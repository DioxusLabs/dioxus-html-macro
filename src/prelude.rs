pub use proc_macro2::TokenStream;

pub use syn::parse::{Parse, ParseBuffer, ParseStream};
pub use syn::{Result, Ident, LitStr, Error, Expr};
pub use quote::{ToTokens, quote};
pub use crate::assertions::*; 
pub use crate::attribute::*; 
pub use crate::attribute_ident::*; 
pub use crate::attributes::*; 
pub use crate::close_tag::*; 
pub use crate::element::*; 
pub use crate::html::*; 
pub use crate::item::*; 
pub use crate::open_tag::*; 
pub use crate::rsx_expr::*; 
