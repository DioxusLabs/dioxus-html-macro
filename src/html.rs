use crate::prelude::*;

#[derive(Default)]
pub struct Html {
    pub elements: Vec<Item>,
}

impl Parse for Html {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let mut elements = vec![];
        while !input.is_empty() && !input.peek2(Token![/]) {
            elements.push(input.parse()?);
        }
        Ok(Html { elements })
    }
}

impl ToTokens for Html {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        for element in &self.elements {
            tokens.extend(quote! {
                #element
            });
        }
    }
}
