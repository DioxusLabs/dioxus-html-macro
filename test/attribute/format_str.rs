use dioxus::prelude::*; 
use dioxus_html_macro::html;

fn main() {
    struct Foo; 

    html!(
        <div "custom"="{Foo}">
        </div>
    );
}
