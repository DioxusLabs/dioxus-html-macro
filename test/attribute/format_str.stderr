error[E0277]: `Foo` doesn't implement `std::fmt::Display`
  --> test/attribute/format_str.rs:7:5
   |
7  | /     html!(
8  | |         <div "custom"="{Foo}">
9  | |         </div>
10 | |     );
   | |_____^ `Foo` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `Foo`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
   = note: this error originates in the macro `format_args` (in Nightly builds, run with -Z macro-backtrace for more info)
